### ISSUE TITLE
[request] Add optional default value to config.get()

### ISSUE THREAD

**Issue by javl** [lorenwest/node-config#504]
### I'm submitting a ... 
  - [x] feature request

I tried finding a request similar to this one in the issuetracker/wiki to see if there is a reason why it's not in config yet but I couldn't find anything. Sorry if I missed it.
I added this request to my local version of the code, but I'm not sure about updating the tests, plus that I wondered if there was any special reason I might be missing why something similar hasn't been implemented yet.

### Request
When getting the value of a property that doesn't exist, instead of catching the exception or working with ```config.has(...)```, wouldn't it be useful to add an optional argument to ```config.get(...)``` to specify a default value to use when the given property doesn't exist?
Something like:

    thing = config.get('name', 'foobar');

If the config contains a ```name``` property, ```thing``` will be set to its value. If ```name``` does not exist, instead of throwing the ```Configuration property 'name' is not defined``` exception, ```thing``` will be set to ```foobar```.

Locally, I updated the get function to look like this (relevant parts only):
```
Config.prototype.get = function(property, defaultValue=null) {
  (...)
  // If the property doesn't exist...
  if (value === undefined) {
    // ... produce an exception if no default value was specified
    if (defaultValue === null) {
      throw new Error('Configuration property "' + property + '" is not defined');
    }
    // or return the given default value
    else {
      value = defaultValue;
    }
  }
};
```
### What is the current behavior?
You can get a value from your config file using ```config.get('name')``` which thows an exception if the property does not exist. You can catch the exception and assign a default value for the non-existing property, or you can do something like:

    if(config.has('name')){
        thing = config.get('name');
    }else{
        thing = 'foobar';
    } 

### Please tell us about your environment:

 -  node-config version: 2.0.1
 -  node-version: 7.10.1

**Comment by lorenwest** [comment]
Hi @javl - this issue has some history, but I'm going to keep this discussion open for a while because I think it's an important topic.

We haven't added this feature because I believe default configurations should be contained in one place. When your servers aren't acting the way they should, at 2am, you want someone (possibly not you) to easily understand the config values running in the application. If they have to sift through tens or hundreds of thousands of lines of code to find default values, it adds time to their task.

We have broken our own rule in this regard by suggesting sub-modules can provide their defaults in code, and I'm not very happy with that solution.

This is one of those "because you can, doesn't mean you should" features. I understand other config libraries offer this ability, but node-config is fairly opinionated.

In short, put your defaults in the default config file. The value of throwing an exception because someone mis-types a config key is far greater than the inconvenience of learning the place to store defaults.

**Comment by markstos** [comment]
@javl See #462 and linked tickets for more history and discussion.

I agree with @lorenwest here. If this feature exists, developers will use it, for better or work. We used to use a configuration system and default values were scattered all over. Some were in the config system. Some were in the code in one place, several were in the code in multiple places.

When all defaults must be in the config system, there's no question where they should be and consistency is maintained. 

The Lodash functions `_.get`, `_.default`, and `_.defaultsTo` all already handle the feature of adding default values directly in code, and they can easily be combined with  `node-config`.

**Comment by javl** [comment]
@lorenwest @markstos Thanks for the replies
Your explanations makes total sense. The whole idea of config is to take some values out of your script of course. This ```default``` argument would work perfectly fine on my current project, as it is fairly small and I'm the only one working on it, but even then I agree it could become unclear.

What I wanted to achieve was to have a way to run my program when certain properties or the config file itself are missing (for whatever reason), but a guess the better approach would be to create a default config file (and disallow modifying that file) together with a second file that will overrule the defaults where needed.

Thanks for taking the time to explain.

**Comment by markstos** [comment]
@javl Yes, using `config/default.js` is the recommended approach.

@lorenwest Should this be closed now?

**Comment by lorenwest** [comment]
Yes, I'll go ahead and close.

**Comment by markstos** [comment]
Publish your `config.getOr()` plugin to NPM and a link to it here: https://github.com/node-config/node-config/wiki/Plugins

**Comment by markstos** [comment]
@peternann Have you published your polyfill?

> My use case is configuration parameters that you ONLY want the code owners to know about. Putting them in config/default.* is the LAST thing you want to do because someone who doesn't know what they are doing might muck with them.

Putting the _actual_ default embedded in the code would help the people who looking at `config/default.js` to *learn what they are doing*, because the _actual_ default value embedded in the code is valuable context. 

The reason that `get()` can thrown exception is so that you can set a value to `undefined` in `config/default.js` but require that a real value is set in environment the app runs.
